[Bryan Seah] iP

bryans17 opened the pull request:
# Duke
> Hi - me 

`Duke` is a todo list, It's:
- text-based
- easy to learn :+1: 
- ~FAST~ **SUPER FAST** to use

To get started,

1. download it from [here](https://github.com/bryans17/ip/releases/tag/A-Jar).
2. double-click it to run.
3. start adding tasks.

Features:
- [x] Add tasks
- [x] Delete tasks
- [x] Mark tasks as done

``` java
public static void main(String[] args) {
        System.out.println("hi");
}
```

Commit:
Add Gradle support

Commit:
welcome message and driver loop

Commit:
Added a Driver Class with add and list functionality

Commit:
Added Task Class and mark/unmark functionality

Commit:
Added new Task subclasses

Commit:
catch ArrayIndexOutOfBoundsException in markAsDone and unmarkDone

Commit:
Added Automated Text UI Testing

Commit:
Created and implemented DukeExceptions

Commit:
Added delete function

Commit:
Added Command Enums and better exception handling

Commit:
Update .gitignore

Commit:
Add TaskList.java

Commit:
Add formatSave() to Task and subclasses

Commit:
Add saveTask() and loadTask()

Commit:
Remove unused imports

Commit:
Update Deadline.java and Event.javato store LocalDateTime

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch-Level-8

Commit:
Update formatSave() in Deadline.java

Commit:
Update variable name deadLine to DEADLINE

Commit:
Add Parser.java

Commit:
Add Ui.java

Commit:
Add Storage.java

Commit:
Update files for OOP

Commit:
Put files into packages

Commit:
Add JUnit Tests

Commit:
Create JAR Manifest

Commit:
Remove unusued Exception Class

Commit:
Add Java Docs

Commit:
Update code to follow coding style standards

Commit:
Add new Command find

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Commit:
Update build.gradle

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Update files to follow checkstyle

Commit:
Update build.gradle

Commit:
Add GUI for Duke using JavaFX

Commit:
Merge branch 'branch-Level-10'

Commit:
Updated code to follow coding standards

LeeEnEn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,72 @@
+public class Parser {
+
+    /**
+     * Parse the command given by the user.
```
Perhaps it should be parses for the method header comment.

bryans17 replied:
Thanks for catching this!

LeeEnEn commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,72 @@
+            return Command.INVALID;
+        }
+    }
+}
```
I like the way you present the parsing of command from the user. Makes the code very easy to follow through. Great work!

LeeEnEn reviewed (commented):
I like the way you comment and format your code, makes it very easy to read and understand. Great work! 👍

bryans17 commented:
> I like the way you comment and format your code, makes it very easy to read and understand. Great work! 👍

Commit:
Add assertions to Duke.java

Currently there are not assertions used in the code.

There is a need to document important assumptions that should hold at various points in the code.

Add assertion in Duke.java to make sure that we are getting a valid
Command enum from Parser.java

Commit:
Update code quality based on SLAP

Code has many areas that can be improved.

Refactor to improve the code quality where necessary.

Implement SLAP principle to multiple files to make code more readable.

Commit:
Use Streams.filter to find tasks with keywords

bryans17 committed:
Merge pull request #1 from bryans17/branch-A-Assertions

Add assertions to Duke.java

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Merge branch 'branch-A-Streams' into branch-A-CodeQuality

bryans17 committed:
Merge pull request #2 from bryans17/branch-A-CodeQuality

Update code quality based on SLAP

Commit:
Add sort Command and feature

bryans17 committed:
Merge pull request #4 from bryans17/branch-C-Sort

Add sort Command and feature

AAlghrairy commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,77 @@
+        case "unmark":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("Missing task number to unmark");
+            }
+            return Command.UNMARK;
+        case "delete":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("Missing task number to delete");
+            }
+            return Command.DELETE;
+        case "todo":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("The description of a todo cannot be empty.");
+            }
+            return Command.TODO;
+        case "deadline":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("The description of a deadline cannot be empty.");
+            }
+            return Command.DEADLINE;
+        case "event":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("The description of an event cannot be empty.");
+            }
+            return Command.EVENT;
+        case "find":
+            if (command.length <= 1 || command[1].equals("")) {
+                throw new DukeWrongInputFormatException("Missing keyword to find");
+            }
+            return Command.FIND;
+        case "sort":
+            if (command.length > 1 && !command[1].equals("")) {
+                throw new DukeWrongInputFormatException("Command sort should not have any arguments");
+            }
+            return Command.SORT;
+        default:
+            return Command.INVALID;
+        }
+    }
+}
```
I think the SLAP coding principle could have been better incorporated here. For instance, rather than having repeated checks for the exception, consider having a separate method that throws an exception if it meets the criteria (e.g. throwIfWrongFormat(command)). This would shorten the method and make it more readable.

AAlghrairy commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,90 @@
+     * @return Formatted String.
+     */
+    private String formatScheduledTime() {
+        DateTimeFormatter form = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
```
Consider abstracting the DateTimeFormatter variables here into static variables in the Task class.

AAlghrairy reviewed (commented):
Looks really great!

Just a few minor issues that have to do with SLAP and encapsulation.

Commit:
Fix minor header comment error

Commit:
Add startup message to be displayed

Commit:
Remove unnecessary line

Commit:
Add more unit tests

Commit:
Add Ui.png

Commit:
Update User Guide

Commit:
Update User Guide

Commit:
Update User Guide

Commit:
Add Quick Start

Commit:
Update minor error

Commit:
Add Ui.png to README
